std/experimental/ndslice/slice.d

```diff
@@ -1737,17 +1738,27 @@ struct Slice(size_t _N, _Range)
             if (this._strides == rslice._strides && this._ptr == rslice._ptr)
                 return true;
         }
-        return opEqualsImpl(this, rslice);
+        foreach (i; Iota!(0, PureN))
+            if (this._lengths[i] == 0)
+                return true;
+        import std.experimental.ndslice.selection : unpack;
+        return opEqualsImpl(this.unpack, rslice.unpack);
     }
 
     ///ditto
     bool opEquals(T)(T[] rarrary)
     {
-        if (this.length != rarrary.length)
+        auto slice = this;
+        if (slice.length != rarrary.length)
             return false;
-        foreach (i, ref e; rarrary)
-            if (e != this[i])
+        if (rarrary.length) do
+        {
+            if (slice.front != rarrary.front)
                 return false;
+            slice.popFront;
+            rarrary.popFront;
+        }
+        while (rarrary.length);
```

Comment:
It is common, because compilers might fail to optimize it. Many compilers *normalize* loops to this form. It gives the invariant that the loop body is executed at least once. This gives a place between `if` and `do` to move loop invariant things out of the loop, but those things will not get executed if the loop is never entered.

I'm not sure if modern compilers still need this manual optimization. It might just be a habit these days.